evm/db/journal.py

```diff
@@ -1,226 +1,258 @@
+
+    #
+    # Database API
+    #
+    def get(self, key: bytes) -> bytes:
+        """
+        For key lookups we need to iterate through the changesets in reverse
+        order, returning from the first one in which the key is present.
+        """
+        # Ignored from mypy because of https://github.com/python/typeshed/issues/2078
+        for changeset_data in reversed(self.journal_data.values()):  # type: ignore
+            if key in changeset_data:
+                return changeset_data[key]
+            else:
+                continue
+
+        return None
+
+    def set(self, key: bytes, value: bytes) -> None:
+        self.latest[key] = value
+
+    def exists(self, key: bytes) -> bool:
+        val = self.get(key)
+        return val is not None and val is not DELETED_ENTRY
 
-    def __contains__(self, value):
-        return value in self.journal_data
+    def delete(self, key: bytes) -> None:
+        self.latest[key] = DELETED_ENTRY
 
 
 class JournalDB(BaseDB):
     """
     A wrapper around the basic DB objects that keeps a journal of all changes.
-    Each time a snapshot is taken, the underlying journal creates a new
-    checkpoint.  The journal then keeps track of the original value for any
-    keys changed.  Reverting to a checkpoint involves merging the original key
-    data from any subsequent checkpoints into the given checkpoint giving
-    precidence earlier checkpoints.  Then the keys from this merged data set
-    are reset to their original values.
+    Each time a recording is started, the underlying journal creates a new
+    changeset and assigns an id to it. The journal then keeps track of all changes
+    that go into this changeset.
+
+    Discarding a changeset simply throws it away inculding all subsequent changesets
+    that may have followed. Commiting a changeset merges the given changeset and all
+    subsequent changesets into the previous changeset giving precidence to later
+    changesets in case of conflicting keys.
+
+    Nothing is written to the underlying db until `persist()` is called.
 
     The added memory footprint for a JournalDB is one key/value stored per
     database key which is changed.  Subsequent changes to the same key within
-    the same checkpoint will not increase the journal size since we only need
-    to track the original value for any given key within any given checkpoint.
+    the same changeset will not increase the journal size since we only need
+    to track latest value for any given key within any given changeset.
     """
     wrapped_db = None
-    journal = None
+    journal = None  # type: Journal
 
-    def __init__(self, wrapped_db):
+    def __init__(self, wrapped_db: BaseDB) -> None:
         self.wrapped_db = wrapped_db
-        self.journal = Journal()
+        self.reset()
 
-    def get(self, key):
-        return self.wrapped_db.get(key)
+    def get(self, key: bytes) -> bytes:
 
-    def set(self, key, value):
+        val = self.journal[key]
```

Comment:
It *appears* that `val` *could* be `DELETED_ENTRY` which would result in returning `DELETED_ENTRY` below.  Am I missing something?

Reply:
Eagle eyes! You're right. But honestly, do we really want to go for this perf improvement? I liked it when the perf improvement seemed to be 5.5x but now that it is down to 1.5 in a *very very* biased benchmark I'm not so crazy about it anymore. Personally I rather prefer the consistency of raising a `KeyError` for missing keys.

Reply:
Yes, that *is* a bug. My comment was about if we maybe want to throw away that entire performance improvement altogether and go back to the API as it was before (the exception ping pong). However, now that I thought about it again I actually think we *should* keep the performance improvement and just fix this bug because in my previous comment I assumed we gave up on the `KeyError` exceptions for this perf improvement but that is actually not the case. The `JournalDB` continues to raise `KeyError` exceptions even with this perf improvement.

Sorry, for the total confusion. I'll just fix the bug and move on haha